mobile-ui/src/main/java/org/sagebionetworks/research/mobile_ui/recorder/device_motion/DeviceMotionRecorder.java

```diff
@@ -0,0 +1,114 @@
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package org.sagebionetworks.research.mobile_ui.recorder.device_motion;
+
+import android.content.Context;
+import android.hardware.SensorManager;
+import android.support.annotation.NonNull;
+
+import com.github.pwittchen.reactivesensors.library.ReactiveSensorEvent;
+
+import org.sagebionetworks.research.mobile_ui.recorder.data.DataLogger;
+import org.sagebionetworks.research.mobile_ui.recorder.ReactiveRecorder;
+import org.sagebionetworks.research.presentation.recorder.DeviceMotionRecorderConfigPresentation;
+
+import java.util.Set;
+
+import io.reactivex.BackpressureStrategy;
+import io.reactivex.Flowable;
+
+/**
+ * A DeviceMotionRecorder recorders a set of the device's motion sensors and provides access to a Single<FileResult>
+ * which can access the data file once the recorder is complete
+ *
+ * @param <S>
+ *         The type of the summary for this recorder.
+ * @param <C>
+ *         The type of the current state for this recorder.
+ */
+public abstract class DeviceMotionRecorder<S, C> extends ReactiveRecorder<S, C, ReactiveSensorEvent> {
+    private static final long MICRO_SECONDS_PER_SEC = 1000000L;
+
+    protected final DeviceMotionSensors deviceMotionSensors;
+
+    protected final double frequency;
+
+    protected final Set<Integer> sensorTypes;
+
+    public DeviceMotionRecorder(DeviceMotionRecorderConfigPresentation config, Context context,
+            DataLogger dataLogger, SummarySubscriber<S, ReactiveSensorEvent> summarySubscriber,
+            CurrentStateSubscriber<C, ReactiveSensorEvent> currentStateSubscriber) {
+        super(config.getIdentifier(), config.getStartStepIdentifier(), config.getStopStepIdentifier(), dataLogger,
+                summarySubscriber, currentStateSubscriber);
+        this.frequency = config.getFrequency();
+        this.sensorTypes = config.getRecorderTypes();
+        int sensorDelay = this.isManualFrequency() ? this.calculateDelayBetweenSamplesInMicroSeconds()
+                : SensorManager.SENSOR_DELAY_FASTEST;
+        this.deviceMotionSensors = new DeviceMotionSensors(context, this.sensorTypes, sensorDelay);
+    }
+
+    @Override
+    public void cancel() {
+        super.cancel();
+        this.deviceMotionSensors.cancel();
+    }
+
+    @Override
+    @NonNull
+    public Flowable<ReactiveSensorEvent> intializeEventFlowable() {
+        return Flowable.create(this.deviceMotionSensors, BackpressureStrategy.BUFFER);
```

Comment:
I haven't thought through all the use cases yet, but it doesn't seem like this does anything to handle when the buffer (default 128) fills up, and consumers would still need to handle backpressure.

Also, I think this creates a cold Flowable and we may want a hot Flowable/Observable, e.g. if a location/distance recorder runs across multiple steps, but only certain steps want to subscribe to the location to display in the UI

Reply:
Yeah I'm still not 100% on all the RxJava constructs and BackPressure is definitely one of the ones I don't know very well. If you want my input on this I can read through some more docs, or if you are sure what we should do I'm happy to defer to you.